config/configcompression/compressiontype.go

```diff
@@ -25,19 +37,45 @@ func (ct *Type) IsCompressed() bool {
 	return *ct != typeEmpty && *ct != typeNone
 }
 
-func (ct *Type) UnmarshalText(in []byte) error {
-	typ := Type(in)
-	if typ == TypeGzip ||
-		typ == TypeZlib ||
-		typ == TypeDeflate ||
-		typ == TypeSnappy ||
-		typ == TypeZstd ||
-		typ == TypeLz4 ||
-		typ == typeNone ||
-		typ == typeEmpty {
-		*ct = typ
+func (ct *TypeWithLevel) UnmarshalText(in []byte) error {
+	var err error
+	parts := strings.Split(string(in), "/")
+	compressionTyp := Type(parts[0])
+	level := zlib.DefaultCompression
+	if len(parts) == 2 {
+		level, err = strconv.Atoi(parts[1])
+		if err != nil {
+			return fmt.Errorf("invalid compression level: %q", parts[1])
+		}
+		if compressionTyp == TypeSnappy ||
+			compressionTyp == TypeLz4 ||
+			compressionTyp == typeNone ||
+			compressionTyp == typeEmpty {
+			return fmt.Errorf("compression level is not supported for %q", compressionTyp)
+		}
+	}
+	ct.Level = Level(level)
+	if (compressionTyp == TypeGzip && isValidLevel(level)) ||
+		(compressionTyp == TypeZlib && isValidLevel(level)) ||
+		(compressionTyp == TypeDeflate && isValidLevel(level)) ||
+		compressionTyp == TypeSnappy ||
+		compressionTyp == TypeLz4 ||
+		compressionTyp == TypeZstd ||
+		compressionTyp == typeNone ||
+		compressionTyp == typeEmpty {
+		ct.Level = Level(level)
+		ct.Type = compressionTyp
 		return nil
 	}
-	return fmt.Errorf("unsupported compression type %q", typ)
 
+	return fmt.Errorf("unsupported compression type and level(default if not specified) %s - %d", compressionTyp, ct.Level)
+
+}
+
+func isValidLevel(level int) bool {
```

Comment:
Should we have zlib/gzip/flate somewhere in the function name?

Reply:
I could add a comment for this function, instead of changing the name.

Reply:
Could do. IMO it would be clearer to include it in the name, but that's just my opinion.

Having said that, I think the function could be inlined with a bit of restructuring in the caller. Something like this:

```go
type TypeWithLevel struct {
        Type  Type   `mapstructure:"type"`

        // Level holds the compression level for algorithms that support one.
        //
        // If Level is nil, then the algorithm's default compression level will be used.
        Level *int `mapstructure:"level"`
}

func (ct *TypeWithLevel) Validate() error {
        switch ct.Type {
        case TypeGzip, TypeZlib, TypeDeflate, TypeSnappy, TypeZstd, TypeLz4, typeNone, typeEmpty:
        default:
                return fmt.Errorf("invalid compression type %q", ct.Type)
        }
        if ct.Level != nil {
                switch ct.Type {
                case TypeGzip, TypeZlib, TypeDeflate:
                        if level < -1 || level > 9 {
                                return fmt.Errorf("invalid compression level %d for %q", *ct.Level, ct.Type)
                        }
                case TypeZstd:
                        // TODO check zstd level
                default:
                        return fmt.Errorf("compression level is not supported for %q", ct.Type)
                }
        }
        return nil
}
```